README
NAME
    MouseX::SimpleConfig - A Mouse role for setting attributes from a simple
    configfile

VERSION
    version 0.10

SYNOPSIS
      ## A YAML configfile named /etc/my_app.yaml:
      foo: bar
      baz: 123

      ## In your class
      package My::App;
      use Mouse;

      with 'MouseX::SimpleConfig';

      has 'foo' => (is => 'ro', isa => 'Str', required => 1);
      has 'baz'  => (is => 'rw', isa => 'Int', required => 1);

      # ... rest of the class here

      ## in your script
      #!/usr/bin/perl

      use My::App;

      my $app = My::App->new_with_config(configfile => '/etc/my_app.yaml');
      # ... rest of the script here

      ####################
      ###### combined with MouseX::Getopt:

      ## In your class
      package My::App;
      use Mouse;

      with 'MouseX::SimpleConfig';
      with 'MouseX::Getopt';

      has 'foo' => (is => 'ro', isa => 'Str', required => 1);
      has 'baz'  => (is => 'rw', isa => 'Int', required => 1);

      # ... rest of the class here

      ## in your script
      #!/usr/bin/perl

      use My::App;

      my $app = My::App->new_with_options();
      # ... rest of the script here

      ## on the command line
      % perl my_app_script.pl -configfile /etc/my_app.yaml -otherthing 123

DESCRIPTION
    This role loads simple configfiles to set object attributes. It is based
    on the abstract role MouseX::ConfigFromFile, and uses Config::Any to
    load your configfile. Config::Any will in turn support any of a variety
    of different config formats, detected by the file extension. See
    Config::Any for more details about supported formats.

    Like all MouseX::ConfigFromFile -derived configfile loaders, this module
    is automatically supported by the MouseX::Getopt role as well, which
    allows specifying "-configfile" on the commandline.

ATTRIBUTES
  configfile
    Provided by the base role MouseX::ConfigFromFile. You can provide a
    default configfile pathname like so:

      has '+configfile' => ( default => '/etc/myapp.yaml' );

    You can pass an array of filenames if you want, but as usual the array
    has to be wrapped in a sub ref.

      has '+configfile' => ( default => sub { [ '/etc/myapp.yaml', '/etc/myapp_local.yml' ] } );

    Config files are trivially merged at the top level, with the right-hand
    files taking precedence.

METHODS
  new_with_config
    Class method provided by the base role MouseX::ConfigFromFile. Acts just
    like regular "new()", but also accepts an argument "configfile" to
    specify the configfile from which to load other attributes. Explicit
    arguments to "new_with_config" will override anything loaded from the
    configfile.

  get_config_from_file
    Class method called internally by either "new_with_config" or
    MouseX::Getopt's "new_with_options". Invokes Config::Any to parse
    "configfile".

BUGS
    Please report any bugs or feature requests on the bugtracker website
    http://github.com/mjg/MouseX-SimpleConfig/issues

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHORS
    *   Brandon L. Black <blblack@gmail.com>

    *   Mark Gardner <mjgardner@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Infinity Interactive.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
0.10      2011-03-05 13:23:37 America/New_York
  - convert release to Dist::Zilla
0.09 - Mar 05, 2011
  - port to Mouse from MooseX::SimpleConfig (mjgardner)

0.09 - Aug 15, 2010
  - Fix issue with MANIFEST

0.08 - Aug 13, 2010
  - Fix passing multiple config files to SimpleConfig.
    - Tests for this (zby)

0.07 - May 22, 2010
  - Reverse order of testing config files so that latter options on the
    command line (when multiple are specified) take precedence. (RT#57027)
  - Reduce exception thrown by being unable to find a config file to a
    warning (RT#57027)
  - Fixed configfile attribute examples
  - Added metadata to Makefile.PL

0.06 - Apr 12, 2010
  - Depend on at least version 0.13 of Config::Any for flatten_to_hash
    support.

0.05 - Jan 22, 2010
   - Allow multiple config files to be loaded and merged.
     Patch in RT#48552 from Sawyer X

0.04 - Nov 04, 2009
    - Allow options to be passed through to Config::Any more easily.

0.03 - Jan 23, 2008
    - Mostly just added pod and tests for configfile
      defaulting

0.02 - Dec 19, 2007
    - Skip the tests if neither of YAML or YAML::Syck is
      installed (but don't make it a test requirement,
      some people won't be using this with YAML at all).

0.01 - Dec 18, 2007
    - Initial release


t/lib/MXSimpleConfigTest.pm
#
# This file is part of MouseX-SimpleConfig
#
# This software is copyright (c) 2011 by Infinity Interactive.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
use utf8;
use strict;
use Modern::Perl;    ## no critic (UselessNoCritic,RequireExplicitPackage)

=encoding utf8

=cut

package MXSimpleConfigTestBase;
use Mouse;

has 'inherited_ro_attr' => ( is => 'ro', isa => 'Str' );

no Mouse;
1;

package MXSimpleConfigTest;
use Mouse;
extends 'MXSimpleConfigTestBase';
with 'MouseX::SimpleConfig';

has 'direct_attr' => ( is => 'ro', isa => 'Int' );

has 'req_attr' => ( is => 'rw', isa => 'Str', required => 1 );

no Mouse;
1;


t/lib/MXDefaultConfigTest.pm
#
# This file is part of MouseX-SimpleConfig
#
# This software is copyright (c) 2011 by Infinity Interactive.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
use utf8;
use strict;
use Modern::Perl;    ## no critic (UselessNoCritic,RequireExplicitPackage)

=encoding utf8

=cut

package MXDefaultConfigTestBase;
use Mouse;

has 'inherited_ro_attr' => ( is => 'ro', isa => 'Str' );

no Mouse;
1;

package MXDefaultConfigTest;
use Mouse;
use Path::Class::File;
extends 'MXDefaultConfigTestBase';
with 'MouseX::SimpleConfig';

has 'direct_attr' => ( is => 'ro', isa => 'Int' );

has 'req_attr' => ( is => 'rw', isa => 'Str', required => 1 );

has '+configfile' => ( default => 'test.yaml' );

no Mouse;
1;


t/lib/MXDriverArgsConfigTest.pm
#
# This file is part of MouseX-SimpleConfig
#
# This software is copyright (c) 2011 by Infinity Interactive.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
use utf8;
use strict;
use Modern::Perl;    ## no critic (UselessNoCritic,RequireExplicitPackage)

=encoding utf8

=cut

package MXDriverArgsConfigTestBase;
use Mouse;

has 'inherited_ro_attr' => ( is => 'ro', isa => 'Str' );

no Mouse;
1;

package MXDriverArgsConfigTest;
use Mouse;
extends 'MXDriverArgsConfigTestBase';
with 'MouseX::SimpleConfig';

has 'direct_attr' => ( is => 'ro', isa => 'Int' );

has 'req_attr' => ( is => 'rw', isa => 'Str', required => 1 );

sub config_any_args {
    return +{ driver_args => { General => { -LowerCaseNames => 1 } } };
}

no Mouse;
1;


lib/MouseX/SimpleConfig.pm
#
# This file is part of MouseX-SimpleConfig
#
# This software is copyright (c) 2011 by Infinity Interactive.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
use utf8;
use strict;
use Modern::Perl;    ## no critic (UselessNoCritic,RequireExplicitPackage)

package MouseX::SimpleConfig;

BEGIN {
    $MouseX::SimpleConfig::VERSION = '0.10';
}

# ABSTRACT: A Mouse role for setting attributes from a simple configfile

use Carp;
use English '-no_match_vars';
use Mouse::Role;
with 'MouseX::ConfigFromFile';

use Config::Any ();

sub get_config_from_file {
    my ( $class, $file ) = @ARG;

    my $files_ref;
    given ( ref $file ) {
        when ('CODE')  { $file      = $file->() }
        when ('ARRAY') { $files_ref = $file }
        default        { $files_ref = [$file] }
    }

    my $can_config_any_args = $class->can('config_any_args');
    my $extra_args
        = $can_config_any_args ? $can_config_any_args->( $class, $file ) : {};
    my $raw_cfany = Config::Any->load_files(
        {   %{$extra_args},
            use_ext         => 1,
            files           => $files_ref,
            flatten_to_hash => 1,
        }
    );

    my %raw_config;
    foreach my $file_tested ( reverse @{$files_ref} ) {
        if ( !exists $raw_cfany->{$file_tested} ) {
            warn qq{Specified configfile '$file_tested' does not exist, }
                . qq{is empty, or is not readable\n};
            next;
        }

        my $cfany_hash = $raw_cfany->{$file_tested};
        croak
            "configfile must represent a hash structure in file: $file_tested"
            if not(    $cfany_hash
                    && ref $cfany_hash
                    && ref $cfany_hash eq 'HASH' );

        %raw_config = ( %raw_config, %{$cfany_hash} );
    }

    return \%raw_config;
}

no Mouse::Role;
1;

__END__

=pod

=for :stopwords Brandon L. Black Mark Gardner Infinity Interactive

=encoding utf8

=head1 NAME

MouseX::SimpleConfig - A Mouse role for setting attributes from a simple configfile

=head1 VERSION

version 0.10

=head1 SYNOPSIS

  ## A YAML configfile named /etc/my_app.yaml:
  foo: bar
  baz: 123

  ## In your class
  package My::App;
  use Mouse;

  with 'MouseX::SimpleConfig';

  has 'foo' => (is => 'ro', isa => 'Str', required => 1);
  has 'baz'  => (is => 'rw', isa => 'Int', required => 1);

  # ... rest of the class here

  ## in your script
  #!/usr/bin/perl

  use My::App;

  my $app = My::App->new_with_config(configfile => '/etc/my_app.yaml');
  # ... rest of the script here

  ####################
  ###### combined with MouseX::Getopt:

  ## In your class
  package My::App;
  use Mouse;

  with 'MouseX::SimpleConfig';
  with 'MouseX::Getopt';

  has 'foo' => (is => 'ro', isa => 'Str', required => 1);
  has 'baz'  => (is => 'rw', isa => 'Int', required => 1);

  # ... rest of the class here

  ## in your script
  #!/usr/bin/perl

  use My::App;

  my $app = My::App->new_with_options();
  # ... rest of the script here

  ## on the command line
  % perl my_app_script.pl -configfile /etc/my_app.yaml -otherthing 123

=head1 DESCRIPTION

This role loads simple configfiles to set object attributes.  It
is based on the abstract role
L<MouseX::ConfigFromFile|MouseX::ConfigFromFile>, and uses
L<Config::Any|Config::Any> to load your configfile.
L<Config::Any|Config::Any> will in
turn support any of a variety of different config formats, detected
by the file extension.  See L<Config::Any|Config::Any> for more details about
supported formats.

Like all L<MouseX::ConfigFromFile|MouseX::ConfigFromFile> -derived configfile
loaders, this
module is automatically supported by the L<MouseX::Getopt|MouseX::Getopt> role
as
well, which allows specifying C<-configfile> on the commandline.

=head1 ATTRIBUTES

=head2 configfile

Provided by the base role L<MouseX::ConfigFromFile|MouseX::ConfigFromFile>.
You can
provide a default configfile pathname like so:

  has '+configfile' => ( default => '/etc/myapp.yaml' );

You can pass an array of filenames if you want, but as usual the array
has to be wrapped in a sub ref.

  has '+configfile' => ( default => sub { [ '/etc/myapp.yaml', '/etc/myapp_local.yml' ] } );

Config files are trivially merged at the top level, with the right-hand files
taking precedence.

=head1 METHODS

=head2 new_with_config

Class method provided by the base role
L<MouseX::ConfigFromFile|MouseX::ConfigFromFile>. Acts just like
regular C<new()>, but also accepts an argument C<configfile> to specify
the configfile from which to load other attributes.  Explicit arguments
to C<new_with_config> will override anything loaded from the configfile.

=head2 get_config_from_file

Class method called internally by either C<new_with_config> or
L<MouseX::Getopt|MouseX::Getopt>'s
C<new_with_options>.  Invokes L<Config::Any|Config::Any> to parse
C<configfile>.

=encoding utf8

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
http://github.com/mjg/MouseX-SimpleConfig/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHORS

=over 4

=item *

Brandon L. Black <blblack@gmail.com>

=item *

Mark Gardner <mjgardner@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Infinity Interactive.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;

BEGIN { require 5.010; }

use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'A Mouse role for setting attributes from a simple configfile',
  'AUTHOR' => 'Brandon L. Black <blblack@gmail.com>, Mark Gardner <mjgardner@cpan.org>',
  'BUILD_REQUIRES' => {
    'File::Find' => '0',
    'File::Temp' => '0',
    'Mouse' => '0',
    'Path::Class::File' => '0',
    'Test::More' => '0.88'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'MouseX-SimpleConfig',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'MouseX::SimpleConfig',
  'PREREQ_PM' => {
    'Carp' => '0',
    'Config::Any' => '0.13',
    'English' => '0',
    'Modern::Perl' => '0',
    'Mouse' => '0.35',
    'Mouse::Role' => '0',
    'MouseX::ConfigFromFile' => '0.02',
    'utf8' => '0'
  },
  'VERSION' => '0.10',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



